### PR TITLE
Convert worker registration to a stream

### PR DESCRIPTION
**Pull request by jiacheliu3**
### What changes are proposed in this pull request?

This change adds the following items:
1. Add an option for the worker register logic to be a streaming call instead of unary
2. Add a stressbench implementation

As a side effect, the locking mechanism in `MasterWorkerInfo` is changed from `ReentrantReadWriteLock` to `StampedLock`. See the discussion in this PR for why.

### Why are the changes needed?

When the worker has a huge amount of blocks to report to the master, the worker register call will be a huge unary RPC call. This incurs a lot of heap allocation on the master side and it's mostly not GC-able before the call finishes. Changing this to streaming makes it more GC-friendly. 

Being able to do some GC during the stream can allow higher concurrency too, so more workers can successfully register at the same time.

Streaming also enables more flow control logic that was impossible when using a unary call. This is not included in this PR but may be added in the future.

### Does this PR introduce any user facing changes?

The stream will hold the write locks on the `MasterWorkInfo` during the stream. As a side effect, the Web UI and other readers of `MasterWorkerInfo` will block when they are inquiring the status of all workers. When the worker finishes registering, or when the stream times out, the locks will be released and the readers will be unblocked. This is one field to improve on in future changes.

### CODE REVIEW

**Comment by xichen01** [comment]
@jiacheliu3 I'm adding some features to the worker register function in this pr:  https://github.com/Alluxio/alluxio/pull/14258 
 I noticed that you had to do some refactoring on this function on the current pr, did I need to sync those changes in my pr now to prevent a mass conflict, or waiting this pr was merged？

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 1]
```diff
@@ -3352,6 +3363,55 @@ public String toString() {
...
+          .setDefaultValue("true")
+          .setDescription("When the worker registers with the master, whether the request should be"
+              + " broken into a stream of smaller batches. This is useful when the worker's storage"
+              + " is large and we expect a large number of blocks. ")
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
+          .setScope(Scope.WORKER)
+          .build();
+  public static final PropertyKey WORKER_REGISTER_STREAM_BATCH_SIZE =
+      new Builder(Name.WORKER_REGISTER_STREAM_BATCH_SIZE)
+          .setDefaultValue("500000")
```
Comment (ggezer) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:3377]:
```suggestion
          .setDefaultValue("1000000")
```

**Review by ggezer — APPROVED** [review]
Thanks for the huge effort Jiacheng!

**Comment by ggezer** [comment]
alluxio-bot, merge this please

**Pull request merged**